lib/core/plugin/plugin.js

```diff
@@ -0,0 +1,318 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2020 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const path = require('path');
+const _ = require('lodash');
+const fs = require('fs');
+
+const PluginContext = require('./context');
+const PrivilegedPluginContext = require('./privilegedContext');
+const kerror = require('../../kerror');
+const Manifest = require('./pluginManifest');
+
+const assertionError = kerror.wrap('plugin', 'assert');
+const runtimeError = kerror.wrap('plugin', 'runtime');
+
+const PLUGIN_NAME_REGEX = /^[a-z-]+$/;
+const HTTP_VERBS = ['get', 'head', 'post', 'put', 'delete', 'patch', 'options'];
+
+class Plugin {
+  constructor (instance, { name, application=false } = {}) {
+    this._instance = instance;
+
+    this._application = application;
+
+    this._initCalled = false;
+    this._config = {};
+    this._context = null;
+    this._version = '';
+    this._name = '';
+    this._manifest = null;
+
+    if (name) {
+      this.name = name;
+    }
+  }
+
+  init (kuzzle, name) {
+    this.name = name;
+
+    if (kuzzle.config.plugins[this.name]) {
+      this.config = JSON.parse(JSON.stringify(kuzzle.config.plugins[this.name]));
+    }
+
+    // check plugin privileged prerequisites
+    // user need to acknowledge privileged mode in plugin configuration
+    if (this.config.privileged) {
+      if (! this.manifest.privileged) {
+        throw assertionError.get('privileged_not_supported', this.name);
+      }
+    }
+    else if (this.manifest && this.manifest.privileged) {
+      throw assertionError.get('privileged_not_set', this.name);
+    }
+
+
+    this._context = this.config.privileged
+      ? new PrivilegedPluginContext(kuzzle, this.name)
+      : new PluginContext(kuzzle, this.name);
+  }
+
+  info () {
+    if (this.application) {
+      return {
+        controllers: this.instance.api,
+        hooks: Object.keys(this.instance.hooks),
+        name: this.name,
+        pipes: Object.keys(this.instance.pipes),
+        version: this.version,
+      };
+    }
+
+    const description = {
+      controllers: [],
+      hooks: [],
+      manifest: this.manifest,
+      pipes: [],
+      routes: [],
+      strategies: [],
+      version: this.version
+    };
+
+    if (_.has(this.instance, 'hooks')) {
```

Comment:
(nitpicking) use `lib/util/safeObject` instead of lodash (much faster methods)
This is a nitpicking because speed isn't relevant here, but using the same methods for all our code is better I think